### PR TITLE
LSP rename integration tests setup

### LINKED ISSUE #61: Add Integration Tests Setup: LSP Renaming

### PR DESCRIPTION
**Pull request by johmsalas**
closes #61

### CODE REVIEW

File: tests/textcase/lsp/lsp_spec.lua [hunk 1]
```diff
@@ -1 +1,62 @@
-describe("LSP", function() end)
+local test_helpers = require("tests.test_helpers")
+local textcase = require("textcase")
+
+local cur_dir = vim.fn.expand("%:p:h")
+
+---@param path string
+function M.read_file(path)
```
Comment (peterfication) [inline tests/textcase/lsp/lsp_spec.lua:7]: I would move this to the test helpers.

File: tests/textcase/lsp/lsp_spec.lua [hunk 2]
```diff
@@ -1 +1,62 @@
...
+  local fd = vim.loop.fs_open(path, "r", 438)
+  local fstat = vim.loop.fs_fstat(fd)
+  local contents = vim.loop.fs_read(fd, fstat.size, 0)
+  vim.loop.fs_close(fd)
+  return contents
+end
+
+function M.waitFor(max_time, callback)
```
Comment (peterfication) [inline tests/textcase/lsp/lsp_spec.lua:15]: This as well. I think the naming convention in Lua is snake_case.

File: tests/textcase/lsp/lsp_spec.lua [hunk 3]
```diff
@@ -1 +1,62 @@
...
+  local max_seconds = max_time
```
Comment (peterfication) [inline tests/textcase/lsp/lsp_spec.lua:16]: I think this is an unnecessary assignment. Just call the function param `max_seconds`.

File: tests/textcase/lsp/lsp_spec.lua [hunk 4]
```diff
@@ -1 +1,62 @@
...
+  local curr_seconds = 0
+  local task_finished = false
+  while curr_seconds < max_seconds and not task_finished do
+    curr_seconds = curr_seconds + 0.5
```
Comment (peterfication) [inline tests/textcase/lsp/lsp_spec.lua:20]: Nitpick: I would probably assign `local step_time = 0.5` outside the function and use it here as well as in `vim.wait(step_time * 1000, function() end)`.

File: tests/textcase/lsp/lsp_spec.lua [hunk 5]
```diff
@@ -1 +1,62 @@
...
+        test_helpers.execute_keys("q")
+        return ts_server_started
+      end)
+
+      assert.is.truthy(ts_server_started)
+      vim.wait(1000, function() end)
+
+      test_helpers.execute_keys("/variableToBeTested<CR>gaS")
+      local content = nil
+      M.waitFor(5000, function()
```
Comment (peterfication) [inline tests/textcase/lsp/lsp_spec.lua:52]: This seems to be a lot. Maybe you changed during development from milliseconds to seconds but missed to change this number to 5?

File: tests/textcase/lsp/lsp_spec.lua [hunk 6]
```diff
@@ -1 +1,62 @@
-describe("LSP", function() end)
+local test_helpers = require("tests.test_helpers")
+local textcase = require("textcase")
+
+local cur_dir = vim.fn.expand("%:p:h")
```
Comment (peterfication) [inline tests/textcase/lsp/lsp_spec.lua:4]: This is not needed. You can just use a relative path for the cmd.

File: tests/textcase/lsp/lsp_spec.lua [hunk 7]
```diff
@@ -1 +1,62 @@
...
+      local ts_server_started = false
+      M.waitFor(30, function()
+        vim.cmd("LspInfo")
+        ts_server_started = not not vim.fn.search("1 client(s) attached to this buffer")
+        test_helpers.execute_keys("q")
+        return ts_server_started
+      end)
+
+      assert.is.truthy(ts_server_started)
+      vim.wait(1000, function() end)
```
Comment (peterfication) [inline tests/textcase/lsp/lsp_spec.lua:48]: 1 second seems a bit long. I think 100ms should be enough.

File: tests/textcase/lsp/lsp_spec.lua (hunk 4 above)
Comment (peterfication) [reply tests/textcase/lsp/lsp_spec.lua:20]: And I would probably try it with 0.1 seconds to speed up the tests.

File: tests/textcase/lsp/lsp_spec.lua [hunk 8]
```diff
@@ -1 +1,62 @@
...
+    it("Should be triggered on keybinding", function()
+      local path = cur_dir .. "/tests/textcase/lsp/fixtures/component-camel-case.tsx"
+      local cmd = " silent exe 'e " .. path .. "'"
+      vim.cmd(cmd)
+      vim.bo.filetype = "typescriptreact"
+      -- allow tsserver start
+
+      local ts_server_started = false
+      M.waitFor(30, function()
+        vim.cmd("LspInfo")
```
Comment (peterfication) [inline tests/textcase/lsp/lsp_spec.lua:41]: Cool idea to check for `LspInfo` :) I was already wondering how you would do that.

**Review by peterfication — DISMISSED** [review]
Really nice to have tests for this now as well! :)
This seems to have been a lot to figure out!

Don't feel obliged to act on every comment ;)

File: tests/textcase/lsp/lsp_spec.lua (hunk 2 above)
Comment (johmsalas) [reply tests/textcase/lsp/lsp_spec.lua:15]: `gaS`

File: tests/textcase/lsp/lsp_spec.lua (hunk 8 above)
Comment (johmsalas) [reply tests/textcase/lsp/lsp_spec.lua:41]: Hopefully it does not become a flaky test. But we do depend on LspInfo Screen

What I'm trying to learn is not to be so perfectionist, but keep progressing and change things later if we find a better way

File: tests/textcase/lsp/lsp_spec.lua (hunk 7 above)
Comment (johmsalas) [reply tests/textcase/lsp/lsp_spec.lua:48]: I wish, even 500ms is not enough. The issue is in the `pipeline` and `act`. `just tests` runs fine with 50ms
I don't even understand what this await is for.

In the previous `wait_for`, it already makes sure the LSP was loaded and it has closed the LspInfo window. So, there is no apparent reason for this `vim.wait`, but if it is missing the pipeline tests does not work

File: tests/textcase/lsp/lsp_spec.lua (hunk 4 above)
Comment (johmsalas) [reply tests/textcase/lsp/lsp_spec.lua:20]: Fair enough

File: tests/textcase/lsp/lsp_spec.lua (hunk 5 above)
Comment (johmsalas) [reply tests/textcase/lsp/lsp_spec.lua:52]: Good catch. Changed to milliseconds.

**Review by johmsalas — COMMENTED** [review]
> I can't leave you all the fun

Have that feeling of accomplishment that all the relevant tests are now setup


> This seems to have been a lot to figure out!

I can't leave you all the fun ;)

File: tests/test_helpers.lua [hunk 9]
```diff
@@ -10,4 +10,23 @@ M.execute_keys = function(feedkeys, mode)
...
+---@param path string
+M.read_file = function(path)
+  local fd = vim.loop.fs_open(path, "r", 438)
+  local fstat = vim.loop.fs_fstat(fd)
+  local contents = vim.loop.fs_read(fd, fstat.size, 0)
+  vim.loop.fs_close(fd)
+  return contents
+end
+
+M.wait_for = function(max_seconds, callback)
```
Comment (johmsalas) [inline tests/test_helpers.lua:22]: fixing, it must be milliseconds

File: tests/textcase/lsp/lsp_spec.lua (hunk 2 above)
Comment (peterfication) [reply tests/textcase/lsp/lsp_spec.lua]: :D 👌

File: tests/textcase/lsp/lsp_spec.lua (hunk 7 above)
Comment (peterfication) [reply tests/textcase/lsp/lsp_spec.lua:26]: Ok :/ good to know! Can you add a comment for this then?

File: tests/textcase/lsp/lsp_spec.lua [hunk 10]
```diff
@@ -1 +1,40 @@
...
+        test_helpers.execute_keys("q")
+        return ts_server_started
+      end)
+
+      assert.is.truthy(ts_server_started)
+      vim.wait(1000, function() end)
+
+      test_helpers.execute_keys("/variableToBeTested<CR>gaS")
+      local content = nil
+      M.wait_for(5 * 1000, function()
```
Comment (peterfication) [inline tests/textcase/lsp/lsp_spec.lua:30]: So this is an interesting one: M.wait_for is not part of this file anymore. However, it still works! And it's calling the method from `test_helpers`. 🤔 Lua has some strange implementations ... I would still change it.

Comment (peterfication) [reply tests/textcase/lsp/lsp_spec.lua:30]: Ah it was me when I created the `test_helpers`. It's missing a `local` before `M = {}`. Can you add this please so this here fails?

Comment (peterfication) [reply tests/textcase/lsp/lsp_spec.lua:30]: Not so strange after all ^^

File: tests/textcase/lsp/lsp_spec.lua (hunk 8 above)
Comment (peterfication) [reply tests/textcase/lsp/lsp_spec.lua:19]: There is `vim.lsp.buf.server_ready()` but it takes way longer.

**Pull request merged**